File: android/app/src/main/kotlin/space/go2china/visepanda/data/model/ButlerModels.kt
package space.go2china.visepanda.data.model

import com.google.gson.annotations.SerializedName

enum class CanvasPatchIntent {
    @SerializedName("create_trip")
    CreateTrip,
    @SerializedName("adjust_trip")
    AdjustTrip,
    @SerializedName("add_alerts")
    AddAlerts,
}

data class AssistantResponse(
    val headline: String,
    val body: String,
    val highlights: List<String>,
    val watchOut: String? = null,
    val nextStep: String,
    // v0.3.12: mirrors the web contract's InlineToolCard[] (lib/types/trip.ts) —
    // see DESIGN.md ADR-116. Rendered as a read-only card in MessageBubble;
    // `href` deep-linking to a Tools category is intentionally not wired up
    // yet because ui/tools/ToolsScreen.kt is still an honest placeholder.
    val toolCards: List<InlineToolCard>? = null,
    // Chat↔Explore bridge (Issue #50/#59): mirrors API_SPEC.md's
    // AssistantResponse.exploreRefs. Server only populates this with POIs the
    // orchestrator actually had real toolContext data for — never rendered as
    // a placeholder when empty/absent.
    val exploreRefs: List<ExploreRef>? = null,
)

data class ExploreRef(
    val amapPoiId: String,
    val name: String,
    val cityId: String,
    val category: ExploreRefCategory,
    val subcategory: String? = null,
    val rating: Double? = null,
    val pricePerPerson: Double? = null,
)

enum class ExploreRefCategory {
    @SerializedName("attractions")
    Attractions,
    @SerializedName("food")
    Food,
    @SerializedName("stays")
    Stays,
}

enum class InlineToolCardTone {
    @SerializedName("info")
    Info,
    @SerializedName("warning")
    Warning,
    @SerializedName("success")
    Success,
}

data class InlineToolCard(
    val id: String,
    val categoryId: String,
    val title: String,
    val summary: String,
    val items: List<String>,
    val nextAction: String,
    val href: String? = null,
    val tone: InlineToolCardTone? = null,
    val sourceLabel: String? = null,
)

data class CanvasPatch(
    val intent: CanvasPatchIntent,
    val assistantMessage: String,
    val assistantResponse: AssistantResponse? = null,
    val tripSummary: TripSummaryPatch? = null,
    val days: List<TripDay>? = null,
    val butlerAlerts: List<ButlerAlert>? = null,
    val reason: String,
)

data class TripSummaryPatch(
    val title: String? = null,
    val durationDays: Int? = null,
    val pace: Pace? = null,
    val travelerStyle: String? = null,
    val destinations: List<String>? = null,
    val confidence: TripConfidence? = null,
)

data class ButlerChatMessage(
    val id: String,
    val role: ButlerMessageRole,
    val content: String,
    val response: AssistantResponse? = null,
    val createdAtEpochMillis: Long,
)

enum class ButlerMessageRole {
    User,
    Assistant,
}

data class ButlerTurnResult(
    val assistantMessage: ButlerChatMessage,
    val trip: TripState,
    val suggestions: List<String>,
    val modelLabel: String,
    val offlineFallback: Boolean,
)

File: ios/VisePandaIOS/Models/ButlerModels.swift
import Foundation

enum CanvasPatchIntent: String, Codable {
    case createTrip = "create_trip"
    case adjustTrip = "adjust_trip"
    case addAlerts = "add_alerts"
}

enum InlineToolCardTone: String, Codable {
    case info
    case warning
    case success
}

struct InlineToolCard: Codable, Identifiable, Equatable {
    var id: String
    var categoryId: String
    var title: String
    var summary: String
    var items: [String]
    var nextAction: String
    var href: String?
    var tone: InlineToolCardTone?
    var sourceLabel: String?
}

struct ButlerExploreRef: Codable, Identifiable, Equatable, Hashable {
    var amapPoiId: String
    var name: String
    var cityId: String
    var category: String
    var subcategory: String?
    var rating: Double?
    var pricePerPerson: String?
    var editorial: Bool?

    var id: String { amapPoiId }

    enum CodingKeys: String, CodingKey {
        case amapPoiId
        case name
        case cityId
        case category
        case subcategory
        case rating
        case pricePerPerson
        case editorial
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        amapPoiId = try container.decode(String.self, forKey: .amapPoiId)
        name = try container.decode(String.self, forKey: .name)
        cityId = try container.decode(String.self, forKey: .cityId)
        category = try container.decode(String.self, forKey: .category)
        subcategory = try container.decodeIfPresent(String.self, forKey: .subcategory)
        rating = try container.decodeFlexibleDouble(forKey: .rating)
        pricePerPerson = try container.decodeFlexibleString(forKey: .pricePerPerson)
        editorial = try container.decodeFlexibleBool(forKey: .editorial)
    }

    init(
        amapPoiId: String,
        name: String,
        cityId: String,
        category: String,
        subcategory: String? = nil,
        rating: Double? = nil,
        pricePerPerson: String? = nil,
        editorial: Bool? = nil
    ) {
        self.amapPoiId = amapPoiId
        self.name = name
        self.cityId = cityId
        self.category = category
        self.subcategory = subcategory
        self.rating = rating
        self.pricePerPerson = pricePerPerson
        self.editorial = editorial
    }
}

private extension KeyedDecodingContainer {
    func decodeFlexibleString(forKey key: Key) throws -> String? {
        if let value = try? decodeIfPresent(String.self, forKey: key) {
            return value.isEmpty ? nil : value
        }
        if let value = try? decodeIfPresent(Double.self, forKey: key) {
            return String(format: "%.0f", value)
        }
        return nil
    }

    func decodeFlexibleDouble(forKey key: Key) throws -> Double? {
        if let value = try? decodeIfPresent(Double.self, forKey: key) {
            return value
        }
        if let value = try? decodeIfPresent(String.self, forKey: key) {
            return Double(value)
        }
        return nil
    }

    func decodeFlexibleBool(forKey key: Key) throws -> Bool? {
        if let value = try? decodeIfPresent(Bool.self, forKey: key) {
            return value
        }
        if let value = try? decodeIfPresent(String.self, forKey: key) {
            return Bool(value)
        }
        return nil
    }
}

struct AssistantResponse: Codable, Equatable {
    var headline: String
    var body: String
    var highlights: [String]
    var watchOut: String?
    var nextStep: String
    var toolCards: [InlineToolCard]?
    var exploreRefs: [ButlerExploreRef]? = nil
}

struct TripSummaryPatch: Codable, Equatable {
    var title: String?
    var durationDays: Int?
    var pace: Pace?
    var travelerStyle: String?
    var destinations: [String]?
    var confidence: TripConfidence?
}

struct CanvasPatch: Codable, Equatable {
    var intent: CanvasPatchIntent
    var assistantMessage: String
    var assistantResponse: AssistantResponse?
    var tripSummary: TripSummaryPatch?
    var days: [TripDay]?
    var butlerAlerts: [ButlerAlert]?
    var affectedDays: [Int]? = nil
    var generationStage: String? = nil
    var reason: String
}

enum ChatRole: String, Codable {
    case user
    case assistant
}

struct ChatMessage: Codable, Identifiable, Equatable {
    var id: String
    var role: ChatRole
    var content: String
    var response: AssistantResponse?
    var affectedDays: [Int]? = nil
    /** Present only on assistant messages that actually changed the trip — see ChangeDigest.swift. */
    var changeDigest: [ChangeDigestEntry]? = nil
    var createdAt: String?
}

struct UserPreferenceProfile: Codable, Equatable {
    var pace: String?
    var budget: String?
    var party: String?
    var dietaryRestrictions: [String]
    var cuisinePreferences: [String]
    var interests: [String]
    var profileConfidence: String
}

struct ToolContext: Codable, Equatable {}

struct ButlerChatRequest: Codable {
    var message: String
    var trip: TripState
    var messages: [ChatMessage]
    var preferenceProfile: UserPreferenceProfile?
    var completeSkeletonFor: TripState? = nil
}

struct ButlerChatResponse: Codable {
    var ok: Bool?
    var fallbackReason: String?
    var mode: String?
    var modelLabel: String?
    var intent: String?
    var strategy: String?
    var providersTried: [String]?
    var patch: CanvasPatch
    var suggestions: [String]?
    var toolContext: ToolContext?
}

struct ButlerTurnResult {
    var assistantMessage: ChatMessage
    var trip: TripState
    var suggestions: [String]
    var modelLabel: String
    var offlineFallback: Bool
}

struct UserMemoryEntry: Codable, Identifiable, Equatable {
    var key: String
    var value: String
    var confidence: Double
    var evidence: [String]
    var source: String
    var updatedAt: String

    var id: String { "\(key)|\(value)|\(updatedAt)" }
}

struct UserMemoryProfileResponse: Codable, Equatable {
    var ok: Bool
    var entries: [UserMemoryEntry]
}

struct UserMemoryDeleteResponse: Codable, Equatable {
    var ok: Bool
    var removed: Bool
}
